### PR TITLE
fix: give specific message instead of only 404 when method disabled

### PR DESCRIPTION
**Pull request by lmcl90**
## Related issue

#954 

## Proposed changes

Return 400 with message `This endpoint was disabled by system administrator. Please check your url or contact the system administrator to enable it.` when user requests endpoints belonging to disabled methods. 

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security
      vulnerability. If this pull request addresses a security. vulnerability, I
      confirm that I got green light (please contact
      [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push
      the changes.
- [x] I have added tests that prove my fix is effective or that my feature
      works.
- [x] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

File: driver/registry_default.go [hunk 1]
```diff
@@ -256,9 +256,7 @@ func (m *RegistryDefault) RegistrationStrategies() registration.Strategies {
 	if len(m.registrationStrategies) == 0 {
 		for _, strategy := range m.selfServiceStrategies() {
 			if s, ok := strategy.(registration.Strategy); ok {
-				if m.c.SelfServiceStrategy(string(s.ID())).Enabled {
-					m.registrationStrategies = append(m.registrationStrategies, s)
-				}
+				m.registrationStrategies = append(m.registrationStrategies, s)
```
Comment (aeneasr) [inline driver/registry_default.go:259]: The problem here is that these strategies are not only used for handlers but also in other areas (e.g. populating the flow methods). I think this needs to be kept as is!

File: driver/registry_default_test.go [hunk 2]
```diff
@@ -68,10 +68,10 @@ func TestDriverDefault_Strategies(t *testing.T) {
 	}{
 		{prep: func(conf *config.Config) {
 			conf.MustSet(config.ViperKeySelfServiceStrategyConfig+".password.enabled", false)
-		}},
+		}, expect: []string{"password", "oidc"}},
 		{prep: func(conf *config.Config) {
 			conf.MustSet(config.ViperKeySelfServiceStrategyConfig+".password.enabled", true)
-		}, expect: []string{"password"}},
+		}, expect: []string{"password", "oidc"}},
```
Comment (aeneasr) [inline driver/registry_default_test.go:74]: The changes in this file should be reverted as this is still the expected behavior!

**Review by aeneasr — CHANGES_REQUESTED** [review]
Awesome, thank you for your contribution! This looks pretty good and I have some ideas how to improve it further :)

Particularly, I think we could use the IsDisabled logic here as well:

https://github.com/ory/kratos/blob/master/driver/registry_default.go#L154-L162

Then, you could change e.g. [this call](https://github.com/ory/kratos/blob/master/driver/registry_default.go#L159) to use a new method such as `GetAllLoginStrategies()` which would simply ignore the `Enabled` property but wouldn't break the existing behavior!

File: driver/registry_default.go (hunk 1 above)
Comment (lmcl90) [reply driver/registry_default.go:259]: Yeah, I see it. I'll refactor this.

Comment (aeneasr) [reply driver/registry_default.go:259]: Awesome, thanks!

File: selfservice/strategy/handler.go [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+const EndpointDisabledMessage = "This endpoint was disabled by system administrator. Please check your url or contact the system administrator to enable it."
+
+func IsDisabled(c interface {
+	config.Provider
+	x.WriterProvider
+}, strategy string, wrap httprouter.Handle) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, ps httprouter.Params) {
+		enabled := c.Config(r.Context()).SelfServiceStrategy(strategy).Enabled
+		if !enabled {
+			c.Writer().WriteError(w, r, herodot.ErrBadRequest.WithError(EndpointDisabledMessage))
```
Comment (aeneasr) [inline selfservice/strategy/handler.go:20]: Let's make this `herodot.ErrNotFound.WithDescription(EndpointDisabledMessage)` :)

**Review by aeneasr — CHANGES_REQUESTED** [review]
Great job! Just one little detail and it's good to go!!

File: selfservice/strategy/handler.go (hunk 3 above)
Comment (lmcl90) [reply selfservice/strategy/handler.go:20]: OK

Comment (lmcl90) [reply selfservice/strategy/handler.go:20]: There is no method named `WithDescription` in `DefaultError`. I think it should be `WithReason`?

**Comment by aeneasr** [comment]
Thank you so much for your work!

**Pull request merged**